### ISSUE TITLE
[no-use-before-define] False positive for mutually recursive functions

### ISSUE THREAD

**Issue by octogonz** [eslint/eslint#12473]
**Tell us about your environment**

* **ESLint Version:** v6.5.1 
* **Node Version:**  v8.15.0
* **npm Version:** v6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

Default

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/*eslint no-use-before-define: error */

function isOdd(num) {
  if (num === 1)
    return true;
  return !isEven(num-1);  // <=== no-use-before-define error
}

function isEven(num) {
  if (num === 0)
    return true;
  return !isOdd(num-1);
}
```

This repros in the [demo page](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdXNlLWJlZm9yZS1kZWZpbmU6IGVycm9yICovXG5cbmZ1bmN0aW9uIGlzT2RkKG51bSkge1xuICBpZiAobnVtID09PSAxKVxuICAgIHJldHVybiB0cnVlO1xuICByZXR1cm4gIWlzRXZlbihudW0tMSk7XG59XG5cbmZ1bmN0aW9uIGlzRXZlbihudW0pIHtcbiAgaWYgKG51bSA9PT0gMClcbiAgICByZXR1cm4gdHJ1ZTtcbiAgcmV0dXJuICFpc09kZChudW0tMSk7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19).

**What did you expect to happen?**

No error should be reported.  The source code is fine.

**What actually happened? Please include the actual, raw output from ESLint.**

`no-use-before-define` reports an error saying `isEven` is used before it is defined.  This "problem" is apparently unsolvable without suppressing ESLint.

**Are you willing to submit a pull request to fix this bug?**

Maybe.

**Comment by g-plane** [comment]
Thanks for this issue.

You may need to configure this rule to disable checks on functions by specifying `functions` to `false`. Detail: https://eslint.org/docs/rules/no-use-before-define#options

**Comment by octogonz** [comment]
> You may need to configure this rule to disable checks on functions by specifying `functions` to `false`.

That's an okay workaround, but it's not a fix.  Functions calling each other is pretty fundamental.

**Comment by g-plane** [comment]
I don't think that's a workaround, since function declarations are hoisted.

**Comment by platinumazure** [comment]
This rule is intentionally designed to be a little more opinionated than most of our rules. It ignores hoisting and ensures that all of your references to a variable occur strictly after the declaration.

It doesn't sound like this is desired behavior for your case. So you have a few options:

1. Turn off function checks, as noted by @g-plane.
1. Restructure your code so the rule passes:
    ```js
    var isOdd, isEven;
    /*eslint no-use-before-define: error */
    
    isOdd = function(num) {
      if (num === 1)
        return true;
      return !isEven(num-1);  // <=== no-use-before-define error
    }
    
    isEven = function(num) {
      if (num === 0)
        return true;
      return !isOdd(num-1);
    }
    ```
1. Disable the rule entirely.

The function configuration referenced by @g-plane is the designated escape hatch for this situation. The rule is working as designed.

**Comment by octogonz** [comment]
> This rule is intentionally designed to be a little more opinionated than most of our rules. It ignores hoisting and ensures that all of your references to a variable occur strictly after the declaration.

In my case we're using TypeScript, so hoisting isn't relevant anyway.  However code is more readable if you declare things before you refer to them, so the rule still seems desirable.  (Compare with TSLint's [no-use-before-declare](https://palantir.github.io/tslint/rules/no-use-before-declare/).

> The rule is working as designed.

It seems that:
- The rule fails for correct, common, reasonable-looking code
- The only possible workaround requires rewriting the code to look worse (or else disabling the check entirely)
- If we improved the design, it wouldn't weaken the protection against hoisting mistakes in any way

If so, is this really the best design?  I might be willing to contribute a PR to improve it (if it's not too much work).  I suppose it would involve detecting cyclic references and exempting them.

**Comment by platinumazure** [comment]
> I suppose it would involve detecting cyclic references and exempting them.

I was going to ask for a concrete proposal on what to change and this might be it.

I could get behind a new option that only detects and exempts cycles.

Let's see what other team members think.

**Comment by g-plane** [comment]
You can only disable or configure the rule for that file, which won't influence the project configuration.

> code is more readable if you declare things before you refer to them

Cyclic references don't look readable to me.

> The rule fails for correct, common, reasonable-looking code

Again, this rule works well. All you need to do is to disable checking for functions.

**Comment by octogonz** [comment]
> You can only disable or configure the rule for that file, which won't influence the project configuration.

We want to minimize suppressions of rules as much as possible. In a large team, when if suppressions become a "normal" practice, it works against the goals of linting.

> Cyclic references don't look readable to me.

None of the proposed workarounds eliminate the cyclic references, though.  My example was pretty artificial, but there are plenty of algorithms that require functions to call each other.  (For example, a recursive descent parser.)
